### DISCUSSION TITLE
Creating ports that wait for connections....

### DISCUSSION THREAD

**Discussion by worikgh** [Boddlnagg/midir#112] [General]
I use a lot of programmes that start up an publish  MIDI ports for input and for output (e.g. Pure Data).

I have no idea how they do it, and I cannot see how to do it with midir.  

How is it done?

**Comment by gcb** [comment]
There might be an issue tracking it, but there's no virtual device support on windows and this doesn't implement custom driver hacks to enable one there.

But on linux/osx, you can use:
```
use midir::os::unix::{VirtualInput, VirtualOutput};                                                                       
use midir::{Ignore, MidiInput, MidiOutput, MidiOutputConnection};
...
let midi_out = MidiOutput::new("MyMidiOutHandle")?;
let mut virt_out = midi_out.create_virtual("My_Visible_MIDI_device_name")?;
```

But I don't know how to watch for connections into it... just blindly send notes there. :shrug:
